peer/upstream/upstream.go

```diff
@@ -0,0 +1,353 @@
+// Copyright (c) 2014-2019 Bitmark Inc.
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package upstream
+
+import (
+	"encoding/binary"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/bitmark-inc/bitmarkd/blockdigest"
+	"github.com/bitmark-inc/bitmarkd/fault"
+	"github.com/bitmark-inc/bitmarkd/mode"
+	"github.com/bitmark-inc/bitmarkd/util"
+	"github.com/bitmark-inc/bitmarkd/zmqutil"
+	"github.com/bitmark-inc/logger"
+	zmq "github.com/pebbe/zmq4"
+)
+
+type UpstreamIntf interface {
+	ActiveInPastSeconds(time.Duration) bool
+	Connect(*util.Connection, []byte) error
+	ConnectedTo() *zmqutil.Connected
+	Destroy()
+	IsConnectedTo([]byte) bool
+	IsConnected() bool
+	GetBlockData(uint64) ([]byte, error)
+	GetClient() zmqutil.ClientIntf
+	RemoteAddr() (string, error)
+	LocalHeight() uint64
+	RemoteDigestOfHeight(uint64) (blockdigest.Digest, error)
+	RemoteHeight() (uint64, error)
+	CachedRemoteHeight() uint64
+	CachedRemoteDigestOfLocalHeight() blockdigest.Digest
+	Name() string
+	ResetServer()
+}
+
+// Upstream - structure to hold an upstream connection
+type Upstream struct {
+	sync.RWMutex
+
+	UpstreamIntf
+
+	log                       *logger.L
+	name                      string
+	client                    zmqutil.ClientIntf
+	connected                 bool
+	remoteHeight              uint64
+	localHeight               uint64
+	remoteDigestOfLocalHeight blockdigest.Digest
+	shutdown                  chan<- struct{}
+	stopPollingSig            chan struct{}
+	lastResponseTime          time.Time
+}
+
+func (u *Upstream) ActiveInPastSeconds(sec time.Duration) bool {
```

Comment:
I think every public function should has comment. At the first look, I cannot understand what it serves for.

Reply:
better code should have less comment

Reply:
totally agree. but `At the first look, I cannot understand what it serves for`. If you are using a library, you dont have document, and you dont have any comments, how do you feel? We also open for community, add comment is necessary. I am a member of Bitmark but I don't understand your code, why others can ?

Reply:
I think name already describe it's function, no need extra comment

Reply:
But please suggest better naming to demonstrate code doing

Reply:
@hxw How do you think?

Reply:
@jamieabc I don't know from where you took this "better code should have less comment", it is totally wrong assumption! In Go if the function can be exported it *MUST* have a comment or otherwise it need to be unexported. 

You should run a golint in your code before commit and open PR!

For your reference: 
"""
Inside a package, any comment immediately preceding a top-level declaration serves as a doc comment for that declaration. Every exported (capitalized) name in a program should have a doc comment. 
"""

URL: https://golang.org/doc/effective_go.html#commentary

Reply:
clean code has some discussion about it

Reply:
briefly speaking, clean code author Martin Flowler thinks code should be self explained, by naming and structure, and not rely on comment to explain

Reply:
Forget about that, because you are not at the level of Martin Flowler, so please, comment your code and follow the best practices, when you reach Martin Flowler's level, you can start to consider to don't document your code, but till there, do so!

Reply:
So, can you give a suggestion about what's missing in the naming?

Reply:
I still believe code should be clean and elegant, so please help to improve the code it's unclear

Reply:
I'm not talking about name, I'm talking about you add a comment on TOP of that exported function to give a brief explanation about that.

Here is how you should do: https://github.com/bitmark-inc/bitmarkd/commit/a0a3ca254322676062c11a33acebac48865a9ba5

@hxw Did pretty much the same over all bitmarkd code, @jamieabc you are not better and you should do the same.

Thank you!

Reply:
@araujobsd OK, all exported functions are having comments

Reply:
@jamieabc Cool! Thanks for that.

Reply:
But I still think naming should be self explained without any comment, so please help me verify those names, I want have a clean method and variable naming.

Reply:
Forget about this "Code as Documentation" thing! Agree code must have good names and be well made and etc, but we are engineers, we should have both: good code and good documentation. Thanks again about the documentation.

Reply:
I agree with Martin Folwer mentioned, although still far way, but I wish someday can have code that is better enough without documentation. But thanks your helping improving the code quality